tests/infra/logging_app.py

```diff
@@ -336,25 +365,97 @@ def get_receipt(self, node, idx, seqno, view, timeout=3, domain="private"):
-                )
-            )
+            url = f"/app/log/{table}?id={log_id}"
+            if self.scope is not None:
+                url += "&scope=" + self.scope
+            check(c.delete(url, headers=self._get_headers_base()))
             if priv:
                 self.priv.pop(log_id)
             else:
                 self.pub.pop(log_id)
 
-    def request(self, log_id, priv=False, log_capture=None):
+    def request(self, log_id, priv=False, log_capture=None, user=None):
         primary, _ = self.network.find_primary(log_capture=log_capture)
-        with primary.client(self.user) as c:
+        with primary.client(user if user else self.user) as c:
             table = "private" if priv else "public"
+            url = f"/app/log/{table}?id={log_id}"
+            if self.scope is not None:
+                url += "&scope=" + self.scope
             return c.get(
-                f"/app/log/{table}?id={log_id}", headers=self._get_headers_base()
+                url,
+                headers=self._get_headers_base(),
+            )
+
+    def post_raw_text(self, log_id, msg, log_capture=None, user=None):
+        primary, _ = self.network.find_primary(log_capture=log_capture)
+        with primary.client(user if user else self.user) as c:
+            url = f"/app/log/private/raw_text/{log_id}"
+            if self.scope is not None:
+                url += "?scope=" + self.scope
+            return c.post(
+                url,
+                msg,
+                headers={"content-type": "text/plain"},
             )
+
+
+def scoped_txs(identity, verify=True):
+    def decorator(func):
+        @functools.wraps(func)
+        def wrapper(*args, **kwargs):
+            network = None
+            node = None
+            previous_txs = None
+            headers = {}
+
+            if isinstance(args[0], infra.network.Network):
+                network = args[0]
+                if hasattr(network, "txs"):
+                    previous_txs = network.txs
+                if hasattr(network, "jwt_issuer") and network.jwt_issuer is not None:
+                    headers = infra.jwt_issuer.make_bearer_header(
+                        network.jwt_issuer.issue_jwt()
+                    )
+                node = network.find_random_node()
+            elif isinstance(args[0], infra.node.Node):
+                node = args[0]
+            else:
+                raise ValueError("invalid args[0]")
+
+            scope = get_fresh_scope(node, identity, headers)
+
+            if network:
+                network.txs = LoggingTxs(identity, scope=scope)
+                network.txs.network = network
+                r = func(*args, **kwargs)
+            else:
+                r = func(*args, **dict(kwargs, scope=scope))
+
+            if network:
+                if verify:
+                    network.txs.verify(network=network)
+                network.txs = previous_txs
```

Comment:
Rather than reconstructing/restoring the entire `LoggingTxs` object, can we just do something like `network.txs.scope = scope` (or push/pop from a scope stack)?

Reply:
Sure, no problem. Might require some changes to the TX verification code, but that shouldn't be too hard to do.